### ISSUE TITLE
Isolation for environments in git based subscriptions

### ISSUE THREAD

**Issue by geowalrus4gh** [akuity/kargo#1370]
# Proposed Feature

Currently, for git-based subscriptions, all the branches including the main branch have all the other environment details committed in.  Is it possible to hold only the environment-specific folder in the respective branch? For example,  It is observed that if we delete 'test' folder from the dev branch, when we promote from dev to test, the 'test' folder gets deleted from the test branch. 

Can we copy the given artifacts only when we promote from branch to branch?

# Motivation

If some environments are to be protected from other environments, say production environment details are to be hidden from the dev and test stages. Dev and test git branches are accessible to the dev team and prod branch is accessible to only the prod team.

# Suggested Implementation

Configure the files to be considered for copying the changes to a target branch,

**Comment by krancour** [comment]
Hi @geowalrus4gh!

There's a lot to this. I'll try to take it bit by bit and see if we can figure something out here.

I'll start with the technical side of things.

> Is it possible to hold only the environment-specific folder in the respective branch?

The tricky thing about this is that for various configuration management tools, a single folder won't be enough. I'll use Kustomize to illustrate...

If you point to a Kustomize overlay in the `test/` dir, for instance, the `kustomization.yaml` file probably references files in a `base/` dir as well. So, right off the bat, there are at least two dirs. Then consider that overlays can be stacked upon overlays. `prod-us-east/` might have region-specific settings on top of `prod/`, which itself is on top of `base/`. Now we're talking about three dirs. The situation for Helm is similar, but involves paths to `values.yaml` files. You might have a base `values.yaml` file and then layers of additional settings/overrides that are used for different stages. And then, of course, we anticipate additional config management tools also being supported via Argo CD-style config management plugins, and can therefore only guess at what the situation looks like for all those unknown tools.

So... how do we get that list of directories that necessary for each stage? There are basically two choices.

1. You tell Kargo what directories are important for each stage
2. Kargo needs to be smart enough to infer which directories are needed for each stage

The problem with no.1 is that it's somewhat onerous for the user.

Their are several problems with no.2:

* _Could_ be done for Kustomize, but wouldn't be easy. Kargo would have to learn how to pick apart a DAG of `kustomization.yaml` files in order to determine all the dirs needed for a stage.
* Could not be done for Helm because your "layers" of `values.yaml` files don't references one another the way that Kustomize overlays do. There is no trail to follow to make inferences about everything that is needed.
* Again, anticipating support for CMPs, there isn't really a strategy for learning all the dirs needed that generalizes for all config management tools.

_There is a third option_ -- just _run_ the config management tool, whichever it may be, because it already knows how to do all of this. Store the _output_ of the config management tool instead of storing the _input_ to the config management tool -- which is what Kargo does otherwise. And, in fact, what I have just described as the best solution to this problem is _exactly_ what Kargo Render does.

So the short and expedient answer to your question is: If this is what you want, use Kargo Render.

If, for some reason, you _don't_ want to use Kargo Render, I think there are some other things worth considering:

> For example, It is observed that if we delete 'test' folder from the dev branch, when we promote from dev to test, the 'test' folder gets deleted from the test branch.

This is definitely not the way things are supposed to work, although I'd consider this behavior to still be within the realm of possibility if you have misconfigured something. If you'd be willing to share the warehouse and stage configurations that you're using when you observe this situation to be occurring, I can see that's going on.

Commenting on this further, however, why or how would the `test/` dir have been deleted from the `dev` branch in the first place?

Can I take a guess that a _person_ has done that? I base that guess on this statement:

    > Dev and test git branches are accessible to the dev team and prod branch is accessible to only the prod team.

Which means there is probably some confusion that needs to be resolved...

Stage/environment-specific branches were not a common pattern before Kargo. In fact, we've heard so many people complain that they've heard this is a bad practice. We've always been fans of it, to be honest, but will concede that in the absence of the right tool, discomfort with it is 100% understandable. We believe Kargo is the tooling that finally makes the practice tenable for most, but a key tenet of that is that those branches are not supposed to be _your_ problem. (We find most people are more accepting of this when they learn this.) They're an implementation detail. They're storage. In a sense, they may as well be S3 buckets instead. (But then this wouldn't be GitOps, right?) What they are not, is they are not supposed to be the user's responsibility to manually maintain.

In summary:

There's a way to do what you want to do and its Kargo Render, but the difficulty you are having probably arises from misconfiguration or from a misunderstanding as to the users' relationships with stage-specific branches, which are purely a means to an end.

Please let me know if this helps at all or if there's anything I'm overlooking here or anything further we could do to help you out.

**Comment by geowalrus4gh** [comment]
First, let me rewrite the statement and avoid the confusion.

>  Dev and test git branches are accessible to the dev team and prod branch is accessible to only the prod team.

to

>  Dev branch is accessible only to dev team. Test branche is accessible only to the test and the prod branch is accessible to only the prod team.

The idea is to, maintain relevant folder only in the respective branches. that is, dev branch should have 'dev' folder only, the test branch should have 'test' folder only and prod branch have 'prod' folder only. 

This ensures the separation of responsibilities. A developer has access to 'dev' branch and he can see the dev configurations. The developer is not allowed to see the test and prod configurations etc. This may vary from case to case though. Isolation is managed using git branches.

> Commenting on this further, however, why or how would the test/ dir have been deleted from the dev branch in the first place?

The use case explained above led us to the deletion of the 'test' folder from the dev branch. We purposely did it. 

> So, right off the bat, there are at least two dirs

Yes, it is ok to have two folders. one for the base folder and one for environment-specific folder. But not other environments' folders. for eg. no 'test' and 'prod' folder in the dev branch. In our use case, we have an umbrella chart, a base values file 'values.yaml' (inside the chart) which will hold all the common values for all environments and another values file 'values-<env>.yaml in the environment folder. 

Please note that, we took an assumption in our use case here. An environment-specific file, ie values-env.yaml, is not supposed to be handled by Kargo and will be managed in respective branches by other means (mostly manually). Changes in the below are only expected to be handled by Kargo promotions

1. Chart templates and other yamls.
2. base values.yaml 
3. Chart.yaml ( This is where we handle versions of dependents charts (apps) to be considered for promotions) 

> Kargo Render

Thanks for this pointer, but I didn't see any docs explaining this feature. Could you please help me to locate if any?

> This is definitely not the way things are supposed to work

This is all what we did. 

1. The initial setup of the application was performed across all environments, involving the initial promotion to each environment. All branches were automatically created from the main branch, and they initially included all environment folders.
2. Subsequently, the 'test' and 'prod' folders were deleted from the main branch.
3. Following that, the 'test' and 'prod' folders were removed from the dev branch.
4. The 'dev' and 'prod' folders were then deleted from the test branch.
5. Similarly, the 'dev' and 'test' folders were removed from the prod branch.

A new freight was introduced and successfully promoted from the 'main' branch to the 'dev' branch. Everything seemed to be in order. However, upon promotion to the 'test' branch, it was observed that the 'test' folder had been deleted.

Apologies for the lengthy description.

**Comment by krancour** [comment]
> Subsequently, the 'test' and 'prod' folders were deleted from the main branch.

Your `main` branch is left with only configuration for dev? It's a bit unconventional to not keep _all_ your config in the trunk.

Kargo executes promotions by taking a specific commit (the commit ID in the freight, which typically is from the trunk), mashing it up with whatever other resources are in the freight (images or charts) and then writing that to _somewhere_ (which happens to be a branch).

Kargo doesn't accommodate any scenario where you're not keeping all your config in the trunk.

**Comment by geowalrus4gh** [comment]
It's okay to have all config in the trunk. Access to the trunk branch can be restricted to the dev team. 
But when I promote from trunk to dev, it will copy all the environment folders to the dev branch as well?

**Comment by krancour** [comment]
> It's okay to have all config in the trunk.

👍 You had me worried for a hot minute.

> But when I promote from trunk to dev, it will copy all the environments folders to the dev branch as well? This is what we wanted to avoid.

Can you elaborate on _why_ you want to avoid this? If those stage-specific branches are viewed as Kargo's internal storage, which I suggested is the right way to think about them, what does it matter?

I'm not saying it _doesn't_ matter, btw. I'm just trying to get to the kernel of the problem.

**Comment by geowalrus4gh** [comment]
In our use case, the dev environment is managed by the development team. They have the privilege to configure the development environment to their choice. So the values in the 'dev' env folder must be configured and committed to the git by the dev team. At the same time, they are restricted from viewing the 'test' and 'prod' configurations. Test and prod configs are managed by the DevOps team.

**Comment by krancour** [comment]
> they are restricted from viewing the 'test' and 'prod' configurations.

Doesn't this contradict your other statement?

> It's okay to have all config in the trunk.

What is the reason for the restriction?

**Comment by geowalrus4gh** [comment]
Sorry for any confusion. 

> It's okay to have all config in the trunk.

This means we can have all the configs in the trunk, provided the development team will be restricted from accessing the trunk branch. They will get access to the dev branch only.

**Comment by krancour** [comment]
Why the restriction?

**Comment by geowalrus4gh** [comment]
There may be sensitive information in the test and production configurations that we [the DevOps team] do not want to expose to the development team. If we give trunk access to the development team and we have all environment configs in the trunk, they are all open to the dev team.

**Comment by krancour** [comment]
It sounds like there is a secrets management problem that you have not addressed. If not addressing it has painted you into a corner where devs and devops cannot work from a single branch, that is the least of your worries. Secrets should not be stored in git.

**Comment by geowalrus4gh** [comment]
It's not only secrets, but the security policy in our organization mandates to hide the production configuration from the dev team. 

For secrets, we have started using sealed secrets like secret management solutions to the stack. Bringing in everything phase by phase. But some dev teams are still using secrets in the git for the time being until they bring in the sealed secrets to stack.

**Comment by krancour** [comment]
> For secrets, we have started using sealed secrets

That is good to know.

> but the security policy in our organization mandates to hide the production configuration from the dev team

What apart from secrets is the policy attempting to protect? All of this, seems like a self-imposed problem.
